Add Home Connect integration

DavidMStraub opened the pull request:
## Description:

This adds an integration to support home appliances (ovens, fridges, hobs, hoods, dryers, washing machines, dish washers) following the Home Connect standard. These are mostly devices by Bosch and related brands (Siemens, Neff, ...).

The integration uses the official REST API (docs) and, after an initial data fetch, works by subscribing to an SSE stream (in a separate thread) for each appliance, i.e. it work by "cloud push".

I developed the underlying Python library `homeconnect` myself (repo here).

The HA integration was already [announced by me](https://community.home-assistant.io/t/home-connect-alpha-testers-wanted/82603) last year in the forums and I received a lot of feedback there and in the [development repository](https://github.com/DavidMStraub/homeassistant-homeconnect/). The custom component is also available via HACS.

While most of the reported issues in the past had to do with the OAuth2 authentication flow, I have now completely rewritten that part using the awesome new official Home Assistant [OAuth2 work flow](https://developers.home-assistant.io/docs/en/api_lib_auth.html) that is also used by Somfy and it works very well.

I feel the integration has now reached a level of maturity and reliability for me to attempt a PR. Having it as an official integration could allow for big improvements in the future:
- It could leverage the new account linking service
- Home Assistant (or Nabu Casa) could apply for enhanced permissions, e.g. switching on ovens remotely, which unfortunately is disabled for normal developer accounts.

Apologies for submitting such a large chunk of code, but it didn't seem reasonable to me to remove parts that are already part of the public custom component.

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io):** https://github.com/home-assistant/home-assistant.io/pull/11336

## Example entry for `configuration.yaml` (if applicable):
```yaml
homeconnect:
  client_id: CLIENT_ID
  client_secret: CLIENT_SECRET
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

DavidMStraub committed:
Initial commit for the Home Connect integrationq

DavidMStraub committed:
Necessary changes in global files

springstan commented on homeassistant/components/homeconnect/manifest.json:
```diff
@@ -0,0 +1,13 @@
+    "@DavidMStraub"
+  ],
+  "requirements": [
+    "homeconnect==0.3.2"
```
I think you forgot to release version `0.3.2` of `homeconnect` before this 😁

DavidMStraub replied:
Oh dear, I don't know what I was doing there :flushed: it should be 0.3.1, actually. Thanks, corrected!

springstan replied:
No problem at all!

DavidMStraub committed:
Correct homeconnect version number

DavidMStraub committed:
Fix unit test for homeconnect config flow

DavidMStraub committed:
Fix pylint error

DavidMStraub committed:
Fix failing unit test

springstan commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,390 @@
+        try:
+            self.appliance.get_status()
+        except (HomeConnectError, ValueError):
+            _LOGGER.debug("Unable to fetch appliance status. Probably offline.")
```
Shouldn't this be a `warning` instead of a `debug` log?
This way a user is not automatically informed when something goes wrong e.g. an exception occurs.

DavidMStraub replied:
This occurs when the appliance is offline during start up, e.g. when the washing machine is switched off, which is a pretty generic/likely scenario, so I wouldn't warn the user about it. Once it comes online (is switched on), the SSE stream should report the necessary updates.

springstan replied:
Okay that sounds very reasonable to me, was just curious why you used debug in these cases.

springstan reviewed (changes requested):
Just some small things to get the reviewing process started :)

springstan commented on homeassistant/components/homeconnect/__init__.py:
```diff
@@ -0,0 +1,114 @@
+        hc_api = data[entry.entry_id]
+        data[DEVICES] = await hass.async_add_executor_job(hc_api.get_devices)
+    except HTTPError as err:
+        _LOGGER.warning("Cannot update devices: %s", err.response.status_code)
```
Only surround the statement which may throw an error, in this case I would assume that is only line 112.

DavidMStraub replied:
Right!

DavidMStraub committed:
Correct documentation links

DavidMStraub committed:
Move import to top level

DavidMStraub committed:
Limit try to single line

springstan reviewed (approved):
Thanks for implementing my suggestions! :)

balloob commented on homeassistant/components/homeconnect/__init__.py:
```diff
@@ -0,0 +1,114 @@
+    return unload_ok
+
+
+@Throttle(SCAN_INTERVAL)
```
For initial contribution this is ok. 

In the future, you might want to put this into a data class that sends a signal when new data is available. Now it can take a while before new data is adopted by entities.

DavidMStraub replied:
Unfortunately the API doesn't push info about new devices. One has to query them and then listen to a separate SSE endpoint for each of them for state updates.

To be honest, I blatantly copied/adapted this function from Somfy without a very deep understanding :grimacing:

balloob commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,389 @@
+        devices = []
+        for app in appl:
+            if app.type == "Dryer":
+                device = Dryer(app)
```
For a future PR, we have a great little helper for this.

```python
from homeassistant.util.decorator import Registry

MODELS = Registry()

@MODELS.register("dryer")
class Dryer(DeviceWithDoor, DeviceWithPrograms):
    …

for app in appl:
    model = MODELS.get(app.type)

    if model is None:
        _LOGGER.warning(…)
        continue

    device = model(app)
    …
```

balloob commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,389 @@
+        """Initialize the device."""
+        self.appliance = appliance
+        try:
+            self.appliance.get_status()
```
We should not do I/O inside a constructor. Add an `initialize` method.

DavidMStraub replied:
I believe this is fixed by https://github.com/DavidMStraub/home-assistant/commit/471d0eb0a0a5684152736dfac06ba09e806721c9 now.

balloob commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,389 @@
+    def get_program_sensors(self):
+        """Get a dictionary with info about program sensors."""
+        sensors = {
+            "Remaining Program Time": "s",
```
We should not have relative times in the state machine. Instead we should use the `timestamp` [device class](https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes) and make the state a UTC timestamp.

balloob replied:
This allows the user to have a countdown in their UI.

BTW, how many programs can be active at the same time? Just 1 right? 

Wouldn't it make more sense to have a start time, stop time, progress % and current program sensor ?

balloob replied:
Or am I misinterpreting something and this is just 1.

DavidMStraub replied:
Right, there is just one remaining time sensor, one progress sensor, and one elapsed time sensor per device/appliance. It's called "sensors" only because there are those 3. I should probably expand the doc string to make it clearer. Basically this maps to what the API gives us.

balloob commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,389 @@
+        """Get a dictionary with info about program sensors."""
+        sensors = {
+            "Remaining Program Time": "s",
+            "Elapsed Program Time": "s",
```
Same.

balloob commented on homeassistant/components/homeconnect/config_flow.py:
```diff
@@ -0,0 +1,28 @@
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow start."""
+        if self.hass.config_entries.async_entries(DOMAIN):
```
Why limit it to 1 installation? It looks like the code can handle multiple config entries ?

DavidMStraub replied:
You're right, I never actually thought about this, I will look into it!

DavidMStraub replied:
Just to be sure I understood correctly: I can just get rid of `async_step_user` altogether and that's it, right? Did that in https://github.com/home-assistant/home-assistant/pull/29214/commits/1b88efb77e83bd60a1c2b65992f7ed4b14af9782.

balloob commented on homeassistant/components/homeconnect/binary_sensor.py:
```diff
@@ -0,0 +1,72 @@
+    def get_entities():
+        entities = []
+        data = hass.data[DOMAIN]
+        for device_dict in data.get(DEVICES, []):
```
This seems like a bug as you write to `hass.data[DOMAIN]conig_entry.entry_id]` in the component.

balloob commented on homeassistant/components/homeconnect/__init__.py:
```diff
@@ -0,0 +1,114 @@
+    data = hass.data[DOMAIN]
+    hc_api = data[entry.entry_id]
+    try:
+        data[DEVICES] = await hass.async_add_executor_job(hc_api.get_devices)
```
We should make a wrapper class that holds `hc_api` and the latest device data, and store that in `hass.data[DOMAIN][entry.entry_id]` so you can support multiple entries.

DavidMStraub replied:
Right!
Since the only other data that I was storing was the list of devices, I simply added it as a property of the existing class now in 
41b6c9d .

balloob commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,389 @@
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Home Connect and the entity name."""
+        return f"{self.device.appliance.haId}-{self.name}"
```
We should just use the haId, as names can change.

balloob replied:
A type would be ok.

DavidMStraub replied:
I had that initially, but then I thought the `unique_id` has to be unique at the entity level, not device level (which is what the `haId` is). Am I misunderstanding?

MartinHjelmare replied:
Bump.

DavidMStraub replied:
OK, just to be sure I understand correctly. At the moment the unique ID has the form, e.g.
`SomeID123-Oven Door`
`SomeID456-Dryer Door`
`SomeID456-Dryer LaundryCare.Dryer.Program.Cotton`

Are you suggesting that I remove the `Oven`/`Dryer` part?

balloob commented on homeassistant/components/homeconnect/sensor.py:
```diff
@@ -0,0 +1,72 @@
+    def icon(self):
+        """Return the icon."""
+        if self._unit == "s":
+            return "mdi:progress-clock"
```
Please add device classes, and when you do, remove the icon as that is taken care by the frontend.

balloob commented on homeassistant/components/homeconnect/switch.py:
```diff
@@ -0,0 +1,185 @@
+
+def format_key(key):
+    """Format Home Connect keys like `BSH.Something.SomeValue` to a simple `some_value`."""
+    if not isinstance(key, str):
```
This seems weird. Why would someone call this without a string ?

balloob replied:
We should move that check outside of this function and put in the call location.

DavidMStraub replied:
I think I did this because I also apply it to the value of the JSON returned by the API (so the function name is a bit of a misnomer) to turn values like `BSH.Common.EnumType.PowerState.On` into `on`; but those values can also be numeric. And I think I wanted to have a single function in order to write a single-line dict comprehension in L185. But that is not strictly necessary of course.

balloob replied:
I would prefer you create a multi-line dict comprehension. Doing this check means we're swallowing errors. I rather have the code blow up when unexpected data comes in. It helps with reusability.

DavidMStraub committed:
Expand doc strings

DavidMStraub committed:
Move device list to ConfigEntryAuth property

DavidMStraub committed:
Include fix to on/off switch
contributed by @MarcusKapun in
https://github.com/DavidMStraub/homeassistant-homeconnect/pull/45

mountainsandcode commented:
Hi @DavidMStraub, is this ready for testing? Happy to give it a case and report back in case it is.

springstan commented:
@mountainsandcode I think you can test it by installing it via HACS as a custom component ✌

balloob commented:
@DavidMStraub what is the status of this PR ?

DavidMStraub commented:
@mountainsandcode @balloob Apologies for the silence and for not having addressed all of the useful comments (thanks for them!). Apart from the holiday break, this was mostly due to an outstanding issue in the underlying library that appears to be causing the SSE stream from the API to break and leaving the devices in a stale state:
https://github.com/DavidMStraub/homeconnect/issues/3

As noted by @springstan, there are thankfully many people testing it and several have reported the problem. I wanted to fix this issue before finalizing the changes to the PR, but haven't found the cause yet. (The SSE code ([source](https://github.com/DavidMStraub/homeconnect/blob/master/homeconnect/sseclient.py)) uses `requests.get(..., stream=True)` under the hood which is supposed to automatically reconnect, and I don't understand why it doesn't. Could it somehow be that Home Assistant closes the thread if no data is received for some time?)

Anyway, I am definitely commited to bringing this to the finish line. I will address the remaining comments in the next 3-4 days and let you know if I manage to fix this.

balloob commented:
`requests.get` does not automatically reconnect.

DavidMStraub commented:
Your're right sorry, I meant keepalive; the requests docs say

> keep-alive is 100% automatic within a session! Any requests that you make within a session will automatically reuse the appropriate connection!

`sseclient` (which I use with minimal changes) then handles reconnects on exceptions
https://github.com/btubbs/sseclient/blob/master/sseclient.py#L97-L100

The problem here seems to be that the integration stops receiving events from the stream after a few hours inspite of this.

DavidMStraub committed:
Bump homeconnect version to 0.3.2

(Attempt to fix the SSE problem)

DavidMStraub committed:
Use homeconnect 0.3.3

DavidMStraub committed:
Move device initialization out of init

DavidMStraub commented:
Sorry, i screwed up a rebase, will fix it now.

DavidMStraub commented:
I fixed the rebase but unfortunately now many people were automatically added as reviewers. I am so sorry for spamming all of you! :frowning_face:

DavidMStraub committed:
isort fixes

DavidMStraub committed:
Bump homeconnect version in requirements*.txt

DavidMStraub committed:
Additional coffee programs (https://github.com/DavidMStraub/homeassistant-homeconnect/pull/60)

DavidMStraub committed:
Bump homeconnect dependency

DavidMStraub commented:
Back to life.

There is hope that I finally fixed the showstopping bug in the library (https://github.com/DavidMStraub/homeconnect/issues/3), so I'm getting back to this. To do list:

- [ ] Confirm that indeed library bug is fixed

and from existing comments:

- [ ] Turn timers into timestamp sensors
- [ ] Allow multiple config entries
- [ ]  add device classes, and when you do, remove the icon as that is taken care by the frontend.
- [ ] Fix `format_key`

DavidMStraub committed:
Fix codespell typos

DavidMStraub committed:
Get rid of format_key

DavidMStraub committed:
Allow multiple config entries

DavidMStraub committed:
Fix pylint error

DavidMStraub committed:
Bump homeconnect to 0.5

DavidMStraub committed:
Convert elapsed/remaining time to timestamp sensors

DavidMStraub committed:
Fix typo

DavidMStraub committed:
Fix sign error

DavidMStraub committed:
Fix JSON lint error

DavidMStraub committed:
Remove elapsed program time sensor, fix for remaining program time

DavidMStraub committed:
Remove title from strings

DavidMStraub committed:
Fix remaining program timestamp sensor

DavidMStraub commented:
@balloob @springstan @frenck 

Thanks again for your useful comments on this PR and again apologies for taking so ridiculously long. But now I am finally confident enough to say that I believe I have addressed all comments and have fixed all bugs reported in the custom integration repo that I would consider show stoppers. And while there are various things still to be improved in the future, **from my side, this PR can be merged now.** :tada:

MartinHjelmare commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,412 @@
+class DeviceWithPrograms(HomeConnectDevice):
+    """Device with programs."""
+
+    _programs = []
```
Can we make this an instance attribute instead? Mutable objects as class attributes are sneaky. Do we really need that?

If this attribute should be considered constant we should show that by writing it in all capital letters.

DavidMStraub replied:
Indeed, they are constants. Capitalized.

MartinHjelmare commented on homeassistant/components/homeconnect/binary_sensor.py:
```diff
@@ -0,0 +1,72 @@
+            entity_dicts = device_dict.get("entities", {}).get("binary_sensor", [])
+            entity_list = [HomeConnectBinarySensor(**d) for d in entity_dicts]
+            device = device_dict["device"]
+            device.entities += entity_list
```
Don't store the entities on the device.

DavidMStraub replied:
Oh right, this is no longer needed now that we're using the dispatcher.

MartinHjelmare commented on homeassistant/components/homeconnect/binary_sensor.py:
```diff
@@ -0,0 +1,72 @@
+    async_add_entities(await hass.async_add_executor_job(get_entities), True)
+
+
+class HomeConnectBinarySensor(HomeConnectEntity, BinarySensorDevice):
```
Please import and use `BinarySensorEntity` instead.

MartinHjelmare commented on homeassistant/components/homeconnect/binary_sensor.py:
```diff
@@ -0,0 +1,72 @@
+        state = self.device.appliance.status.get("BSH.Common.Status.DoorState", {})
+        if not state:
+            self._state = None
+        elif state.get("value", None) in [
```
`None` is the default value returned if the key is missing in the dict.

MartinHjelmare commented on homeassistant/components/homeconnect/config_flow.py:
```diff
@@ -0,0 +1,22 @@
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class OAuth2FlowHandler(config_entry_oauth2_flow.AbstractOAuth2FlowHandler):
```
Pass the domain as keyword argument domain and remove the decorator.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#defining-your-config-flow

MartinHjelmare commented on homeassistant/components/homeconnect/manifest.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "domain": "homeconnect",
```
Our standard is to separate words with underscore. If the name of the brand is Home Connect we should name the integration `"home_connect"`.

DavidMStraub replied:
Oh ... OK.

MartinHjelmare commented on homeassistant/components/homeconnect/sensor.py:
```diff
@@ -0,0 +1,97 @@
+                    self._state = None
+                else:
+                    seconds = self._sign * float(status[self._key]["value"])
+                    self._state = dt_util.utcnow() + timedelta(seconds=seconds)
```
The timestamp sensor should use an absolute utc isoformatted datetime as state.

DavidMStraub replied:
Oh, good to know. Is this actually documented anywhere?

MartinHjelmare commented on homeassistant/components/homeconnect/switch.py:
```diff
@@ -0,0 +1,189 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ["homeconnect"]
```
Please remove `DEPENDENCIES`.

MartinHjelmare commented on homeassistant/components/homeconnect/switch.py:
```diff
@@ -0,0 +1,189 @@
+    async_add_entities(await hass.async_add_executor_job(get_entities), True)
+
+
+class HomeConnectProgramSwitch(HomeConnectEntity, SwitchDevice):
```
Please use `SwitchEntity` instead of `SwitchDevice`.

MartinHjelmare commented on homeassistant/components/homeconnect/switch.py:
```diff
@@ -0,0 +1,189 @@
+
+    def turn_on(self, **kwargs):
+        """Start the program."""
+        _LOGGER.debug("tried to turn on program %s", self.program_name)
```
Please start logging messages with capital letter.

MartinHjelmare commented on homeassistant/components/homeconnect/switch.py:
```diff
@@ -0,0 +1,189 @@
+            self.device.appliance.start_program(self.program_name)
+        except HomeConnectError as err:
+            _LOGGER.error("Error while trying to start program: %s", err)
+        self.async_entity_update()
```
We can't call an async function from a sync context. Then we need to use `hass.add_job` to schedule the async work from the sync context.

Or change `turn_on` to the async version and schedule the sync work on the executor thread pool with `hass.async_add_executor_job`.

It's generally better to have the outer context be async and schedule inner things on the executor as needed, while trying to do that in as large chunks as possible to minimize context switching.

DavidMStraub replied:
OK thanks for the hint, can you help me out a little more since I'm not good at async?
- If I start the program with `async_add_executor_job` and `await` that, is the `async_entity_update` is guaranteed to be called *after* the job has been executed? (This is crucial as the state of the switch needs to be updated after making the API call.)
- Does it make sense to have the `await hass.async_add_executor_job` inside the `try`...`except` or how do I otherwise log the traceback of the sync API call?

MartinHjelmare replied:
1. Yes.
2. Yes, the try... except should wrap the executor job, but nothing more.

MartinHjelmare commented on homeassistant/components/homeconnect/switch.py:
```diff
@@ -0,0 +1,189 @@
+        elif self.device.appliance.status.get(
+            "BSH.Common.Status.OperationState", {}
+        ).get("value", None) in [
+            "BSH.Common.EnumType.OperationState.Ready",
```
Can we extract the constant lists of strings to module level constants so we don't need to create them every update?

DavidMStraub replied:
You're absolute right, will do so.

MartinHjelmare commented on homeassistant/components/homeconnect/switch.py:
```diff
@@ -0,0 +1,189 @@
+                newval = val
+            else:
+                raise ValueError("Unexpected type for value")
+            attributes[newkey] = newval
```
We prefer to set attribute keys explicitly so we know what they are and can check formatting.

Can we create a dictionary that maps the attribute keys to the status keys, and look up that dictionary using the status key before setting the state attributes here?

DavidMStraub replied:
The problem is that I don't really know which attributes are reported by different types of devices and it was frequently requested by users of the custom integration to get access to some attribute (oven temperature, hood light status, stuff like that) for which there is no specific sensor, so it seemed to me to make sense to add all of them so users can create template sensors for whatever they need.

MartinHjelmare replied:
Normally we want measurements exposed as individual sensor entities. That's also a reason why we want to review all state attributes. If we don't know what the measurements are we can not add them.

DavidMStraub replied:
Yes, that makes sense. Adding more sensors is something for the future, but I currently cannot realistically research and test all the relevant measurements. So should I remove the attributes altogether then for the time being?

MartinHjelmare replied:
Yeah, that's best for now. :+1:

DavidMStraub commented on homeassistant/components/homeconnect/__init__.py:
```diff
@@ -0,0 +1,113 @@
+from . import api, config_flow
+from .const import DOMAIN, OAUTH2_AUTHORIZE, OAUTH2_TOKEN
+
+os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = "1"
```
Oh, good catch, I think it is. It was required originally for reverse proxy setups ([issue](https://github.com/DavidMStraub/homeassistant-homeconnect/issues/3)) but indeed with the HA built-in OAuth integration this should definitely removed.

DavidMStraub committed:
Minor changes

DavidMStraub committed:
Move HomeConnectEntity to separate module

DavidMStraub committed:
Remove state attributes
https://github.com/home-assistant/core/pull/29214#discussion_r418968535

DavidMStraub committed:
Make turn_on/off async

DavidMStraub committed:
Put reused string constants into .const

DavidMStraub committed:
Use dispatcher for updating entities

DavidMStraub commented on homeassistant/components/homeconnect/api.py:
```diff
@@ -0,0 +1,412 @@
+        _LOGGER.debug(self.entities)
+        _LOGGER.debug(self.appliance.status)
+        for entity in self.entities:
+            entity.async_entity_update()
```
I believe this is implemented now in https://github.com/DavidMStraub/home-assistant/commit/f7a4dac8803b37945b2d01ab325cc56e3398a00d. Could you have a critical look at that commit? Thanks!

MartinHjelmare commented on homeassistant/components/homeconnect/binary_sensor.py:
```diff
@@ -0,0 +1,72 @@
+        """Return true if the binary sensor is available."""
+        return self._state is not None
+
+    def update(self):
```
Does this method do I/O or could it be run in the event loop?

DavidMStraub replied:
Right, it doesn't do I/O. I can just to `async def async_update` without needing to `await` anything, right?

MartinHjelmare replied:
Yes! :+1:

DavidMStraub committed:
Update homeassistant/components/homeconnect/entity.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub committed:
Update homeassistant/components/homeconnect/entity.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub committed:
Update homeassistant/components/homeconnect/entity.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub commented on homeassistant/components/homeconnect/entity.py:
```diff
@@ -0,0 +1,64 @@
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Home Connect and the entity name."""
+        return f"{self.device.appliance.haId}-{self.name}"
```
OK.

DavidMStraub replied:
Done in 6d9d626f16cd4daf1329784da6423451f225a426

DavidMStraub committed:
Make door entity update async

DavidMStraub committed:
Do not use device name in unique ID

MartinHjelmare commented on homeassistant/components/homeconnect/entity.py:
```diff
@@ -0,0 +1,71 @@
+    @property
+    def unique_id(self):
+        """Return the unique id base on the id returned by Home Connect and the entity name."""
+        return f"{self.device.appliance.haId}-{self.desc}"
```
If `self.desc` is `None` we shouldn't add it. When can it be `None`? Is it just for the power switch? Should we come up with a fake `desc` for that switch to make it future proof if we add more switches in the future?

DavidMStraub replied:
True, that's stupid. I simply use `desc="Power"` now, it's anyway a more appropriate name.

DavidMStraub committed:
Rename power switch

DavidMStraub commented:
@MartinHjelmare please disregard that I just re-requested a review. There have been requested changes that I had overlooked. Will need to adress them first.

DavidMStraub committed:
Rename method

DavidMStraub committed:
Capitalize class attribute

DavidMStraub committed:
Use BinarySensorEntity

DavidMStraub committed:
Explicit None not needed

DavidMStraub committed:
Unneeded decorator

DavidMStraub committed:
Rename integration to home_connect

DavidMStraub committed:
Use isoformat for timestamp sensor

DavidMStraub committed:
Remove unneeded DEPENDENCIES

DavidMStraub committed:
Do not store entities on device

DavidMStraub committed:
Update requirements_test_all.txt

DavidMStraub committed:
Rename also test directory

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -0,0 +1,372 @@
+        {"name": "Dishcare.Dishwasher.Program.MaximumCleaning"},
+    ]
+
+    def get_entit(self):
```
Typo.

DavidMStraub replied:
:scream:

DavidMStraub committed:
Typo

MartinHjelmare commented on homeassistant/components/home_connect/sensor.py:
```diff
@@ -0,0 +1,92 @@
+                        dt_util.utcnow() + timedelta(seconds=seconds)
+                    ).isoformat()
+            else:
+                self._state = status[self._key].get("value", None)
```
```suggestion
                self._state = status[self._key].get("value")
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -0,0 +1,158 @@
+    def update(self):
+        """Update the switch's status."""
+        state = self.device.appliance.status.get(BSH_ACTIVE_PROGRAM, {})
+        if state.get("value", None) == self.program_name:
```
```suggestion
        if state.get("value") == self.program_name:
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -0,0 +1,158 @@
+    def update(self):
+        """Update the switch's status."""
+        if (
+            self.device.appliance.status.get(BSH_POWER_STATE, {}).get("value", None)
```
```suggestion
            self.device.appliance.status.get(BSH_POWER_STATE, {}).get("value")
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -0,0 +1,158 @@
+        ):
+            self._state = True
+        elif (
+            self.device.appliance.status.get(BSH_POWER_STATE, {}).get("value", None)
```
```suggestion
            self.device.appliance.status.get(BSH_POWER_STATE, {}).get("value")
```

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -0,0 +1,158 @@
+        ]:
+            self._state = True
+        elif (
+            self.device.appliance.status.get(BSH_OPERATION_STATE, {}).get("value", None)
```
```suggestion
            self.device.appliance.status.get(BSH_OPERATION_STATE, {}).get("value")
```

MartinHjelmare commented on homeassistant/components/home_connect/sensor.py:
```diff
@@ -0,0 +1,92 @@
+        """Return true if the sensor is available."""
+        return self._state is not None
+
+    def update(self):
```
Does this method do I/O?

DavidMStraub replied:
No, making async ...

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -0,0 +1,158 @@
+            _LOGGER.error("Error while trying to stop program: %s", err)
+        self.async_entity_update()
+
+    def update(self):
```
Does this method do I/O?

DavidMStraub replied:
No, making async ...

MartinHjelmare commented on homeassistant/components/home_connect/switch.py:
```diff
@@ -0,0 +1,158 @@
+            self._state = True
+        self.async_entity_update()
+
+    def update(self):
```
I/O?

DavidMStraub replied:
No, making async ...

MartinHjelmare commented on homeassistant/components/home_connect/binary_sensor.py:
```diff
@@ -0,0 +1,65 @@
+from .entity import HomeConnectEntity
+
+_LOGGER = logging.getLogger(__name__)
+
```
We should set `PARELLEL_UPDATES` constant to limit or not limit number of parallel updates for all entities within this platform. We should do this for all platforms.

https://developers.home-assistant.io/docs/integration_fetching_data#request-parallelism

DavidMStraub replied:
Now that all entities have `async_update`, is this still necessary? Of course the SSE threads are parallel, but this is unrelated to `PARELLEL_UPDATES` I suppose.

MartinHjelmare replied:
If there's no I/O in the update and we use `async_update` we don't need to set parallel updates. It will default to no limit.

A future PR could refactor and drop the use of `async_update` and instead do the processing inside `_update_callback` and call `self.async_write_ha_state` at the end. This would remove the extra scheduling on the event loop that we don't need.

DavidMStraub committed:
Update homeassistant/components/home_connect/sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub committed:
Update homeassistant/components/home_connect/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub committed:
Update homeassistant/components/home_connect/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub committed:
Update homeassistant/components/home_connect/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub committed:
Update homeassistant/components/home_connect/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

DavidMStraub committed:
Make update methods async

MartinHjelmare reviewed (approved):
Awesome!
